### ISSUE TITLE
[sklearn#34128] [META] scalability problems on hybrid CPUs

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#350]
**Mirrored from [scikit-learn#34128](https://github.com/scikit-learn/scikit-learn/issues/34128)**

Original title: [META] scalability problems on hybrid CPUs
Original labels: Parallelism and scalability
Created: 2026-05-26T15:02:50Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

Modern CPUs such as Intel Core Ultra Series 3 processors and Apple M1/M2/... series have different core speeds (P-cores, E-cores, LP-cores...).

joblib and native libraries such as OpenMP or BLAS libraries will often start one thread per core. However, thread scheduling overhead can cause performance degradation compared to using fewer threads on those architectures, especially when iterating on many small data problems, for instance when bagging or grid searching models on small data.

The purpose of this issue is to assess how frequent and how bad is this problem in practice and track coordinate remediation efforts.

## Concrete cases of this problem

- https://github.com/scikit-learn/scikit-learn/issues/30662
- https://github.com/scikit-learn/scikit-learn/issues/17334#issuecomment-4545158514
- ... (comment or edit to add more)

## Possible remediation strategies

- Refactor parallel sections to group more work per parallel task (possibly dynamically based on the data size)
- Dynamically adjust the number of threads based on a cost model informed by CPU architecture, data shape or benchmark: this sounds complex
- Limit the number of threads on hybrid GPUs (e.g. either 1 thread per performance core, or 1 thread per performance core and 1 additional thread per two efficiency core for instance):
  - https://github.com/joblib/loky/pull/630

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [68](https://github.com/godofecht/flow-scikit/issues/68) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
